[신고하기] 신고하기 비즈니스 로직 및 API 구성(#19)

Minjae-An opened the pull request:
## 작업 대상

<!-- 작업한 대상을 자세히 설명해주세요. -->

- 신고 엔티티(`Report`)
- 신고 repository(`ReportRepository` )
- 신고 command(`ReportCommand`)
- 신고하기 비즈니스 로직(`ReportService` )
- 신고하기 API(`UserController` )

## 📄 작업 내용

<!-- 작업한 내용을 간단히 요약해주세요. -->

- 신고 엔티티 `content` 칼럼, `reason` 으로 이름 수정
- 신고 엔티티의 생성/수정/삭제 작업을 수행하는 `Repository` 정의
- 신고하기 비즈니스 로직 구현
    1. 신고 요청 검증
        - 신고 분류가 기타일 때 사유가 존재하는 지 검증
        - 사유가 존재할 때 최대 길이(1024자)를 넘지 않는 지 검증
    2. 자기 자신을 신고하는 예외 상황 검증
    3. 신고한 사용자, 신고 당한 사용자 조회
    4. 신고 엔티티 생성 및 저장
- 신고하기 API 구현

## 🙋🏻 주의 사항

<!-- 리뷰어를 위해 복잡하거나 중요한 코드를 명시해주세요. -->

## 📎 관련 이슈

로직 정상 작동 여부 검증중 request body가 존재하지 않을 때 발생하는 `HttpMessageNotReadableException`이 `ExceptionAdvice` 에서 잡히지 않는 문제 발생, 추후 수정예정.

## 레퍼런스

<!-- 작업 시 참고했던 문건의 URL을 남겨주세요 -->

Minjae-An committed:
feat: 신고 command repository 추가(#19)

신고 엔티티에 대한 생성/수정/삭제 작업을 수행하는 repository 생성

Minjae-An committed:
feat: 신고 관련 enum 정의(#19)

다음 예외 상황에 대한 enum을 정의하였다.
- 신고 분류가 기타인데 신고 사유가 존재하지 않을 경우
- 신고 사유가 최대 길이(1024자)를 초고할 경우
- 자기 자신을 신고하는 경우

Minjae-An committed:
feat: 신고 엔티티 수정(#19)

- 카테고리 필드 not null 조건 정의
- 필드 content, 이름 reason으로 수정
- JpaAuditing 활용, createdAt 필드 엔티티 저장시 자동 설정

Minjae-An committed:
feat: 신고 생성 요청시 카테고리 제약 조건 설정(#19)

- not null, 공백만으로 구성 불가
- 최대 길이 255자

Minjae-An committed:
feat: 신고하기 비즈니스 로직 구성(#19)

신고하기는 다음 과정을 거쳐 이뤄진다.

1. 신고 요청 검증
- 카테고리가 기타가 아닐 경우 사유는 null이므로 검증하지 않음
- 카테고리가 기타인데 사유가 존재하지 않는 경우 예외 발생
- 사유 존재시 길이가 최대 길이(1024자)를 초과할 경우 예외 발생

2. 자기 자신을 신고하는 상황 검증
3. 신고한 유저, 신고 당한 유저 조회
4. 신고 엔티티 생성 및 저장

Minjae-An committed:
feat: 신고 하기 API 구현(#19)

acceptor-gyu commented on be/src/main/java/yeonba/be/mypage/entity/Report.java:
```diff
@@ -32,8 +36,22 @@ public class Report {
+    @CreatedDate
     private LocalDateTime createdAt;
 
+    public Report(
+        User user,
+        User reportedUser,
+        String category,
+        String reason) {
```
controller에서만 파라미터에 어노테이션이 많이 붙을 수 있기 때문에 한 줄에 하나의 파라미터를 쓰는 거를 생각했는데 소통이 부족했던 것같네요. 이 부분에 대해서 정하면 좋을 것같습니다. 지금 이렇게 하면 제가 봤을 때는 가독성이 떨어지는 것같아서요.

Minjae-An replied:
- 파라미터 길이, 개수에 따라 한 줄 한 파라미터 컨벤션을 유동적으로 적용
- 일괄적으로 한 줄 한 파라미터 컨벤션을 적용


크게 두 방향이 있다고 보는데 개인적으로는 일단 일괄적으로 적용하고 추후에 가독성 개선을
위해 리팩토링을 진행하는 게 낫지 않나 싶습니다.

acceptor-gyu replied:
네 그렇게 하면 될 것같습니다. 저도 일단 2번째 방식으로 하겠습니다.

acceptor-gyu commented on be/src/main/java/yeonba/be/user/controller/UserController.java:
```diff
@@ -105,22 +111,22 @@ public ResponseEntity<CustomResponse<Void>> deleteFavorite(
+  @ApiResponse(responseCode = "202", description = "신고 정상 처리")
   @PostMapping("/users/{userId}/report")
   public ResponseEntity<CustomResponse<Void>> report(
+      @RequestAttribute("userId") long userId,
       @Parameter(description = "신고 대상 사용자 ID", example = "1")
-      @PathVariable long userId,
+      @PathVariable("userId") long reportedUserId,
       @RequestBody UserReportRequest request) {
```
controller에서 크게 RequestBody, PathVariable, RequestAttribute 이 3가지 어노테이션이 사용되는데 이 순서도 맞추는 것이 좋을 것같네요. 이후 PR들을 보니 RequestAttribute -> PathVariable -> RequestBody 순서로 작성하신 것같은데 그 이유가 있나요?

Minjae-An replied:
서버에서 설정한 파라미터가 클라이언트가 요청한 파라미터보다 우선순위를 가진다고 생각하여 
가장 먼저 기입했고, path variable과 request body의 경우 '특정 리소스(path variable)에 임의의 
데이터(request body)를 포함해 요청한다'는 흐름에 따라 순서를 정했습니다. 
이 부분은 의견 나누고 협의하면 좋을 것 같습니다.

acceptor-gyu replied:
`우선순위를 가진다`에서 우선순위의 기준이 무엇인가요?

Minjae-An replied:
서버가 리소스에 대한 제어권을 가지고 있기에 서버가 설정한 파라미터가 리소스를 요청하는 클라이언트가 설정한 파라미터보다 우선권을 가져야 한다고 생각했습니다.

acceptor-gyu reviewed (approved):
Comment 확인부탁드립니다. 큰 문제는 없어서 approve 하겠습니다.

acceptor-gyu commented on be/src/main/java/yeonba/be/mypage/repository/ReportCommandRepository.java:
```diff
@@ -0,0 +1,10 @@
+@Repository
+public interface ReportCommandRepository extends JpaRepository<Report, Long> {
+
+}
```
Repository를 이렇게 나눴을 때 2개의 Repository가 하나의 엔티티를 관리하게 되는데 문제가 없을까요? 제가 잘 몰라서 알려주시면 감사하겠습니다.
저는 Repository는 하나만 있고 Command, Query는 하나의 Repository가 갖고 있는 메서드를 사용하는 거에 대해서는 어떻게 생각하시나요?

Minjae-An committed:
refactor: 사용하지 않는 신고 command repository 삭제(#19)

service와 repository를 중개하는 별도의 query,command 계층 구성에
따라 사용하지 않는 command repository 삭제

Minjae-An committed:
refactor: 불필요한 생성자 삭제(#19)

역직렬화에 불필요한 AllArgsConstructor 삭제

Minjae-An committed:
feat: 신고 repository 정의(#19)

Minjae-An committed:
feat: 신고 저장 기능 구현(#19)

Minjae-An committed:
refactor: 신고 command에 의존하도록 로직 수정(#19)

Minjae-An committed:
refactor: 필요 이상으로 분리된 검증 메서드들 단순 작성(#19)

- 카테고리가 기타일 때 사유가 존재하지 않는 경우
- 사유가 최대 길이(1024자)를 초과하는 경우

Minjae-An committed:
Merge branch 'dev' into feat/#19-make-report

Minjae-An committed:
refactor: 자기 자신을 신고하는 상황 검증 로직 수정(#19)

자기 자신을 신고하는 상황을 User의 동일 객체 비교 로직을 활용해
검증토록 수정

Minjae-An committed:
refactor: 코드 정렬(#19)

Minjae-An committed:
refactor: 사용하지 않은 예외 enum 삭제(#19)

Minjae-An committed:
feat: 예외 enum 공통 메서드 정의(#19)

예외 enum들에서 공통적으로 구현해야 하는 메서드들
인터페이스로 정의

Minjae-An committed:
feat: 일반 예외 enum 정의(#19)

서비스 예외가 아닌 HTTP에서 발생할 수 있는 일반적 예외에 대한 enum 정의

Minjae-An committed:
feat: 사용자 관련 예외 enum 정의 및 사용위치 수정(#19)

사용자 엔티티와 관련하여 서비스에서 고유하게 발생할 수 있는 예외 enum 정의.
기존 예외 enum 사용 위치 수정

Minjae-An committed:
feat: 신고 관련 예외 enum 정의(#19)

Minjae-An committed:
refactor: 예외 인터페이스 사용해 예외 정보 제공토록 수정(#19)

예외 enum들이 공통적으로 구현하는 BaseException 인터페이스를
통해 예외 관련 정보를 제공하도록 코드 수정

Minjae-An committed:
refactor: 새로 정의한 신고 예외 enum 사용토록 코드 수정(#19)

Minjae-An committed:
refactor: 새로 정의한 일반 예외 enum 사용토록 코드 수정(#19)
